lib/clients/rtm/client.js

```diff
@@ -505,33 +515,106 @@ RTMClient.prototype.sendMessage = function sendMessage(text, channelId, optCb) {
  * @param {Function=} optCb
  */
 RTMClient.prototype.send = function send(message, optCb) {
-  var wsMsg = cloneDeep(message);
-  var jsonMessage;
   var err;
-  var _this = this;
+  var ret;
 
   if (this.connected) {
-    wsMsg.id = this.nextMessageId();
-    jsonMessage = JSON.stringify(wsMsg);
-    this.logger('debug', jsonMessage);
+    ret = this._send(message, optCb);
+  } else {
+    err = new SlackRTMSendError('ws not connected, unable to send message', message);
+    this.logger('error', err);
 
-    this._pendingMessages[wsMsg.id] = wsMsg;
-    this.ws.send(jsonMessage, undefined, function handleWsMsgResponse(wsRespErr) {
-      if (!isUndefined(wsRespErr)) {
-        _this.logger('debug', 'Unable to send message: ' + wsRespErr);
-      }
+    if (optCb) {
+      setTimeout(function wsSendDisconnectedCb() {
+        optCb(err, null);
+      }, 1);
+    } else {
+      ret = new Promise(function wsSendDisconnectedPromise(fulfill, reject) {
+        reject(err);
+      });
+    }
+  }
+
+  return ret;
+};
 
-      if (!isUndefined(optCb)) {
-        optCb(wsRespErr);
+
+/**
+ *
+ * @param message
+ * @param optCb
+ * @returns {*}
+ * @private
+ */
+RTMClient.prototype._send = function _send(message, optCb) {
+  var _this = this;
+  var jsonMessage;
+  var ret;
+  var msgId = this.nextMessageId();
+
+  var wsMsg = cloneDeep(message);
+  wsMsg.id = msgId;
+
+  jsonMessage = JSON.stringify(wsMsg);
+  this.logger('debug', 'sending message via ws: ' + jsonMessage);
+
+  if (optCb) {
+    _this._registerMsgHandler(msgId, optCb);
+    this.ws.send(jsonMessage, undefined, function handleWsMsgResponseCb(wsRespErr) {
+      if (!isUndefined(wsRespErr)) {
+        _this._handleMsgResponse(msgId, wsRespErr);
       }
     });
   } else {
-    err = 'ws not connected, unable to send message';
-    this.logger('debug', err);
-    if (!isUndefined(optCb)) {
-      optCb(new Error(err));
+    ret = new Promise(function wsSendPromiseInner(fulfill, reject) {
+      _this._registerMsgHandler(msgId, { fulfill: fulfill, reject: reject });
+      _this.ws.send(jsonMessage, undefined, function handleWsMsgResponseCb(wsRespErr) {
+        if (!isUndefined(wsRespErr)) {
```

Comment:
This is the callback for the websocket's send function. The reason we don't call the callback for success here is that we're treating message success as being when the server responds to us with a message object containing a `reply_to` key and a `ts` key.